src/support/multipart-form-data.js

```diff
@@ -0,0 +1,126 @@
+        // Future performance enhancement, should issues arise: keep the _buffer_ in memory
+        // and stream the data elsewhere (e.g. to S3) later on.
+        fileBuffer = Buffer.concat([fileBuffer, data]);
+      }).on('close', () => {
+        // Reached the end of the file data
+        requestData[name] = fileBuffer.toString('utf8');
+      }).on('limit', () => {
+        // File exceeded the size limit
+        reject(new Error(`File size limit exceeded: ${maxFileSizeMb}MB`));
+      });
+    });
+
+    // Handle other fields as JSON, falling back to a string on error
+    busboy.on('field', (name, val) => {
+      let parsedValue;
+      try {
+        parsedValue = JSON.parse(val);
+      } catch (error) {
+        // Not valid JSON, treat as string
+        parsedValue = val;
+      }
+      requestData[name] = parsedValue;
+    });
+
+    // Handle the end of the request
+    busboy.on('close', () => {
+      resolve(requestData);
+    });
+
+    busboy.on('error', (error) => {
+      reject(new Error(`Invalid request: ${error.message}`));
+    });
+
+    request.body.pipe(busboy);
+  });
+}
+
+export function isMultipartFormData(headers) {
+  // Check if the Content-Type header exists and starts with "multipart/form-data"
+  const contentType = headers['content-type'] || headers['Content-Type'];
+  return contentType && contentType.startsWith('multipart/form-data');
+}
+
+/**
+ * Wraps a function with a multipart/form-data middleware that extracts the request data. Only acts
+ * on requests with the multipart/form-data Content-Type.
+ *
+ * @param {UniversalFunction} func - The universal function
+ * @returns {UniversalFunction} A universal function with the added middleware
+ */
+export function multipartFormData(func) {
+  return async (request, context) => {
+    // Only act on requests which use the multipart/form-data Content-Type
+    const { pathInfo: { headers } } = context;
+    if (isMultipartFormData(headers) && !isObject(context.multipartFormData)) {
+      const {
+        MULTIPART_FORM_FILE_COUNT_LIMIT = 1, // Default to a max of 1 file upload
+        MULTIPART_FORM_MAX_FILE_SIZE_MB = 20, // Defaults to a 20MB max, per file
+      } = context.env;
+      try {
+        // Parse the request body and store it in the context
+        context.multipartFormData = await getData(
+          request,
+          headers,
+          MULTIPART_FORM_FILE_COUNT_LIMIT,
+          MULTIPART_FORM_MAX_FILE_SIZE_MB,
+        );
+      } catch (e) {
+        const { log = console } = context;
+        const message = `Error parsing request body: ${e.message}`;
+        log.error(message, e);
+        return new Response('', {
+          status: 400,
+          headers: {
+            'x-error': message,
+          },
+        });
+      }
+      // Return a new request, following existing pattern of the body-data-wrapper
+      return func(new Request(request.url, request.init), context);
```

Comment:
@bhellema, note that this middleware wrapper was directly inspired by https://github.com/adobe/helix-shared/blob/main/packages/helix-shared-body-data/src/body-data-wrapper.js#L101 and I tried to match the existing style, with eventual intention to propose a PR to this repo.